model/flow/identity.go

```diff
@@ -398,7 +398,7 @@ func (il IdentityList) PublicStakingKeys() []crypto.PublicKey {
 }
 
 func (il IdentityList) Fingerprint() Identifier {
-	return MerkleRoot(GetIDs(il)...)
+	return GetIDs(il).Fingerprint()
```

Comment:
> I think we need to use the Checksum field here.

I think the problem is the function is meant for purposes besides fingerprinting, as specified by `fingerprint.Fingerprinter`, yet is confusingly similar to this interface in name and declaration:
* `IdentityList.Fingerprint` is used to uniquely identify a set of identities, by node ID
* `fingerprint.Fingerprinter` is used to generate an encoding of an entity

It would be inappropriate to use `Checksum` over full identities to identify a set of node IDs, because elements of the identity may change over time. (Now, they are static over the course of an epoch, but looking forward, they may change block-to-block). For example, the cluster ID is the hash over all the node IDs of the cluster participants -- it would be inappropriate if a member of the cluster being ejected mid-epoch caused the cluster ID to change. 

I'd instead like to change the name of `IdentityList.Fingerprint` and keep the behaviour the same. https://github.com/onflow/flow-go/pull/3903/commits/12f0668d25eba1b146d2e91a87790212caf4a334

Reply:
I think we agree on the premise:
> elements of the identity may change over time

But I feel we subtly draw different conclusions:
* You are saying that we have a specific use-case (i.e. generate a cluster ID). Therefore, you shape the API to satisfy the use case. 
* My standpoint is: most importantly, the API needs to promote _safe_ usage by default (which I think is the area where your proposal https://github.com/onflow/flow-go/commit/12f0668d25eba1b146d2e91a87790212caf4a334 could be improved):
   * Pretty soon (part of the dynamic identity table workstream), we want to track the potentially changing identities block by block in a fork-aware manner. Meaning, we need a concept of "this checksum represents _exactly_ the dynamic identities as of block X". 
   * Furthermore, `StaticID` is the _only_ method available. Therefore, the "easiest" for an engineer, that is in a rush or unaware of potential security risks, is to use the methods available. You see? The default path of least effort leads to potential security vulnerabilities that are _very_ subtle to spot. 

   Therefore, in my opinion, if `IdentityList` only provides a _single_ method for generating an ID of sort, it must be the cryptographically safe variant (the one I suggested above).

    * Regarding your point:
       >  it would be inappropriate if a member of the cluster being ejected mid-epoch caused the cluster ID to change
      
       Totally agree. Nevertheless, I believe this is a concern for the code generating the cluster ID, i.e. the code that _interacts_ with `IdentityList`. But from the perpective of 


I believe for a security-first system like Flow, 





🤔  I have a subtly different view:
*

Reply:
The reason why I am going into detail here is because I want to promote “safe-by-design” approach for our APIs. The default, minimal-knowledge-required usage of an API should be as safe as possible. In this particular case, the security considerations are quite subtle but severe comparing the different approaches:
1. Assume the IdentityList.Fingerprint commits to the full identities including variable fields.
    - For cases, where we need a commitment to the full dynamic identities, this method is safe.   
    - As you pointed out, it would be buggy to use Fingerprint for generating the cluster ID, because a collector node being ejected would change the ID of the respective cluster. Lets assume we made that bug. Implications: The cluster ID, which the happy path logic assumes to be constant throughout an Epoch, could erroneously change. Thereby, it would probably break the happy path, i.e. undermine liveness of the respective cluster. 
2. Lets now consider the alternative design, where IdentityList.StaticID only commits to the node IDs.
    - If a rushed engineer erroneously used this method in a case where a commitment to the full dynamic identities would be required, we have a huge problem: Essentially byzantine nodes can modify the public keys of any nodes, and still generate a consistent identifier for the identity list.
    - On the positive side, generating cluster ID becomes very simple, without risk for accidental bugs. 


Comparing approach 1. with 2., I feel there is a very important different:
* Approach 1, potentially undermines liveness of the network by breaking the happy path. Though, protocol safety is probably maintained. Presumably, the bug will be relatively apparent and straightforward to find the root cause. 
* In comparison, approach 2 does not impact the happy path. Instead it undermines protocol safety by opening a security vulnerability.

Reply:
**Suggestion**
I think `IdentityList` should implement  the `Entity` interface. It's a broadly-known interface and the ID method is quite consistent with `Identity.ID()` as the latter also returned the node ID. 

Could you please include a explicit warning 
```golang
// CAUTION: 
//  * An IdentityList's ID is a cryptographic commitment to only node IDs. A node operator
//    can freely choose the ID for their node. There is no relationship whatsoever between the node ID
//    and the node's keys. 
//  * To generate a cryptographic commitment for the full IdentityList, use method `Checksum()`. 
//  * The outputs of `IdentityList.ID()` and `IdentityList.Checksum()` are both order-sensitive. 
//    Therefore, the `IdentityList` canonical order , unless explicitly specified otherwise, by the protocol. 
```

Reply:
Good point, I agree. Applied the suggestion here: https://github.com/onflow/flow-go/pull/3903/commits/7ac4e1ca9805b6131cb658863622d6158d6a9983